File: backend/src/main/java/cha_revelacao/controller/EventoController.java
package cha_revelacao.controller;

import cha_revelacao.dto.request.EventoRequest;
import cha_revelacao.dto.response.EventoResponse;
import cha_revelacao.model.Evento;
import cha_revelacao.service.EventoService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/eventos")
public class EventoController {

    private final EventoService eventoService;

    public EventoController(EventoService eventoService) {
        this.eventoService = eventoService;
    }

    @PostMapping
    public ResponseEntity<EventoResponse> criar(@Valid @RequestBody EventoRequest request,
                                               Authentication authentication) {
        EventoResponse response = eventoService.criarEvento(request, authentication.getName());
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping
    public ResponseEntity<List<EventoResponse>> listarTodos() {
        return ResponseEntity.ok(eventoService.listarTodos());
    }

    @GetMapping("/ativos")
    public ResponseEntity<List<EventoResponse>> listarEventosAtivos() {
        return ResponseEntity.ok(eventoService.listarEventosAtivos());
    }

    @GetMapping("/votacao-aberta")
    public ResponseEntity<List<EventoResponse>> listarEventosComVotacaoAberta() {
        return ResponseEntity.ok(eventoService.listarEventosComVotacaoAberta());
    }

    @GetMapping("/meus-eventos")
    public ResponseEntity<List<EventoResponse>> listarEventosDoUsuario(Authentication authentication) {
        return ResponseEntity.ok(eventoService.listarEventosDoUsuario(authentication.getName()));
    }

    @GetMapping("/{id}")
    public ResponseEntity<EventoResponse> buscarPorId(@PathVariable Long id) {
        return ResponseEntity.ok(eventoService.buscarPorId(id));
    }

    @PutMapping("/{id}")
    public ResponseEntity<EventoResponse> atualizar(@PathVariable Long id,
                                                  @Valid @RequestBody EventoRequest request,
                                                  Authentication authentication) {
        return ResponseEntity.ok(eventoService.atualizarEvento(id, request, authentication.getName()));
    }

    @PutMapping("/{id}/revelar")
    public ResponseEntity<EventoResponse> revelarResultado(@PathVariable Long id,
                                                        @RequestParam Evento.SexoBebe resultado,
                                                        Authentication authentication) {
        return ResponseEntity.ok(eventoService.revelarResultado(id, resultado, authentication.getName()));
    }

    @PutMapping("/{id}/encerrar-votacao")
    public ResponseEntity<EventoResponse> encerrarVotacao(@PathVariable Long id,
                                                       Authentication authentication) {
        return ResponseEntity.ok(eventoService.encerrarVotacao(id, authentication.getName()));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deletar(@PathVariable Long id, Authentication authentication) {
        eventoService.deletarEvento(id, authentication.getName());
        return ResponseEntity.noContent().build();
    }
}

File: backend/src/Controllers/EventoController.cs
namespace MemuVie.Evento.Controllers;

using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using MemuVie.Evento.DTOs.Requests;
using MemuVie.Evento.DTOs.Responses;
using MemuVie.Evento.Models;
using MemuVie.Evento.Services;
using System.Security.Claims;

[ApiController]
[Route("/api/eventos")]
public class EventoController : ControllerBase
{
    private readonly IEventoService _eventoService;
    private readonly ILogger<EventoController> _logger;

    public EventoController(IEventoService eventoService, ILogger<EventoController> logger)
    {
        _eventoService = eventoService;
        _logger = logger;
    }

    [HttpPost]
    [Authorize]
    public async Task<ActionResult<EventoResponse>> Criar([FromBody] EventoRequest request)
    {
        try
        {
            var email = User.FindFirst(ClaimTypes.Email)?.Value;
            if (string.IsNullOrEmpty(email))
            {
                return Unauthorized(new ApiResponse(false, "Usuário não autenticado"));
            }

            var response = await _eventoService.CriarEventoAsync(request, email);
            return StatusCode(StatusCodes.Status201Created, response);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao criar evento");
            return BadRequest(new ApiResponse(false, "Erro ao criar evento"));
        }
    }

    [HttpGet]
    public async Task<ActionResult<IEnumerable<EventoResponse>>> ListarTodos()
    {
        try
        {
            var eventos = await _eventoService.ListarTodosAsync();
            return Ok(eventos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao listar eventos");
            return BadRequest(new ApiResponse(false, "Erro ao listar eventos"));
        }
    }

    [HttpGet("ativos")]
    public async Task<ActionResult<IEnumerable<EventoResponse>>> ListarEventosAtivos()
    {
        try
        {
            var eventos = await _eventoService.ListarEventosAtivosAsync();
            return Ok(eventos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao listar eventos ativos");
            return BadRequest(new ApiResponse(false, "Erro ao listar eventos ativos"));
        }
    }

    [HttpGet("votacao-aberta")]
    public async Task<ActionResult<IEnumerable<EventoResponse>>> ListarEventosComVotacaoAberta()
    {
        try
        {
            var eventos = await _eventoService.ListarEventosComVotacaoAbertaAsync();
            return Ok(eventos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao listar eventos com votação aberta");
            return BadRequest(new ApiResponse(false, "Erro ao listar eventos com votação aberta"));
        }
    }

    [HttpGet("meus-eventos")]
    [Authorize]
    public async Task<ActionResult<IEnumerable<EventoResponse>>> ListarEventosDoUsuario()
    {
        try
        {
            var email = User.FindFirst(ClaimTypes.Email)?.Value;
            if (string.IsNullOrEmpty(email))
            {
                return Unauthorized(new ApiResponse(false, "Usuário não autenticado"));
            }

            var eventos = await _eventoService.ListarEventosPorUsuarioAsync(email);
            return Ok(eventos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao listar eventos do usuário");
            return BadRequest(new ApiResponse(false, "Erro ao listar eventos do usuário"));
        }
    }

    [HttpGet("{id}")]
    public async Task<ActionResult<EventoResponse>> BuscarPorId(long id)
    {
        try
        {
            var evento = await _eventoService.BuscarPorIdAsync(id);
            return Ok(evento);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao buscar evento");
            return NotFound(new ApiResponse(false, "Evento não encontrado"));
        }
    }

    [HttpPut("{id}")]
    [Authorize]
    public async Task<ActionResult<EventoResponse>> Atualizar(long id, [FromBody] EventoRequest request)
    {
        try
        {
            var email = User.FindFirst(ClaimTypes.Email)?.Value;
            if (string.IsNullOrEmpty(email))
            {
                return Unauthorized(new ApiResponse(false, "Usuário não autenticado"));
            }

            var evento = await _eventoService.AtualizarAsync(id, request, email);
            return Ok(evento);
        }
        catch (UnauthorizedAccessException ex)
        {
            return Unauthorized(new ApiResponse(false, ex.Message));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao atualizar evento");
            return BadRequest(new ApiResponse(false, "Erro ao atualizar evento"));
        }
    }

    [HttpDelete("{id}")]
    [Authorize]
    public async Task<IActionResult> Deletar(long id)
    {
        try
        {
            var email = User.FindFirst(ClaimTypes.Email)?.Value;
            if (string.IsNullOrEmpty(email))
            {
                return Unauthorized(new ApiResponse(false, "Usuário não autenticado"));
            }

            await _eventoService.DeletarAsync(id, email);
            return NoContent();
        }
        catch (UnauthorizedAccessException ex)
        {
            return Unauthorized(new ApiResponse(false, ex.Message));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao deletar evento");
            return NotFound(new ApiResponse(false, "Evento não encontrado"));
        }
    }

    [HttpPost("{id}/encerrar-votacao")]
    [Authorize]
    public async Task<IActionResult> EncerrarVotacao(long id)
    {
        try
        {
            await _eventoService.EncerrarVotacaoAsync(id);
            return NoContent();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao encerrar votação");
            return NotFound(new ApiResponse(false, "Evento não encontrado"));
        }
    }

    [HttpPost("{id}/revelar")]
    [Authorize]
    public async Task<IActionResult> Revelar(long id, [FromBody] BabySex resultado)
    {
        try
        {
            await _eventoService.RevelarAsync(id, resultado);
            return NoContent();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao revelar resultado");
            return NotFound(new ApiResponse(false, "Evento não encontrado"));
        }
    }
}
